sip/sip-006-runtime-cost-assessment.md

```diff
@@ -0,0 +1,625 @@
+binding the arguments to their passed values, and the cost of
+ensuring that those arguments are of the correct type. Type checks
+and argument binding are _linear_ in the size of the arguments.
+
+The cost of applying a function is:
+
+
+```
+(a*X+b) + costEval(body)
+```
+
+where a and b are constants,
+X := the cumulative size of the argument types,
+costEval(body) := the cost of executing the body of the function
+
+### Type Parsing
+
+Parsing a type in Clarity incurs a linear cost in the size of the
+AST describing the type:
+
+```
+type_parsing_cost(X) = (a*X+b)
+```
+
+where a, b, are constants,
+X := the number of elements in the type description AST
+
+### Function Definition
+
+Defining a function in Clarity incurs an execution cost at the
+time of contract publishing (unrelated to any analysis). This
+is the cost of _parsing_ the function's signature, which is linear
+in the length of the type signatures, and linear in the length of the
+function name and argument names.
+
+```
+binding_cost + sum(type_parsing_cost(Y) + a, ARG_TYPES)
+```
+
+where
+ARG_TYPES := the length of the argument type signatures
+and a is a constant associated with the binding of argument types.
+
+### Contract Storage Cost
+
+Storing a contract incurs both a runtime cost as well as storage costs. Both of
+these are _linear_ the size of the contract AST.
+
+```
+WRITE_LENGTH = a*X+b
+RUNTIME_COST = c*X+d
+```
+
+where a, b, c, and d, are constants.
+
+# Native Function Costs
+
+## Data, Token, Contract-Calls ##
+
+### Data Lookup Costs
+
+Fetching data from the datastore requires hashing the key to be looked up.
+That cost is linear in the key size:
+
+```
+data_hash_cost(X) = a*X+b
+```
+
+X := size of the key
+
+### Data Fetching Costs
+
+Fetching data from the datastore incurs a runtime cost, in addition to
+any costs associated with MARF accesses (which are simply counted as the
+integer number of times the MARF is accessed). That runtime cost
+is _linear_ in the size of the fetched value (due to parsing).
+
+```
+read_data_cost = a*X+b
+```
+
+X := size of the fetched value.
+
+### Data Writing Costs
+
+Writing data from the datastore incurs a runtime cost, in addition to
+any costs associated with MARF writes (which are simply counted as the
+integer number of times the MARF is written). That runtime cost
+is _linear_ in the size of the written value (due to data serialization).
+
+```
+write_data_cost = a*X+b
+```
+
+X := size of the stored value.
+
+### contract-call
+
+Contract calls incur the cost of a normal function lookup and
+application, plus the cost of loading that contract into memory from
+the data store (which is linear in the size of the called contract).
+
+```
+RUNTIME_COST: (a*Y+b) + func_lookup_apply_eval(X)
+READ_LENGTH: Y
+```
+
+where a and b are constants,
+Y := called contract size
+`func_lookup_apply_eval(X)` := the cost of looking up, applying, and
+evaluating the body of the function
```

Comment:
Does this include the cost of looking up the trait definition(s) that the contract-call may invoke?

Reply:
This PR doesn't deal with traits yet (since the traits PR itself hasn't finalized yet).